File: libnd4j/include/ops/declarable/generic/sparse/coo_to_csr.cpp
/* ******************************************************************************
 *
 *
 * This program and the accompanying materials are made available under the
 * terms of the Apache License, Version 2.0 which is available at
 * https://www.apache.org/licenses/LICENSE-2.0.
 *
 *  See the NOTICE file distributed with this work for additional
 *  information regarding copyright ownership.
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 * License for the specific language governing permissions and limitations
 * under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 ******************************************************************************/

//
// COO → CSR sparse conversion op
// Inputs: [0] indices [nnz, 2] INT (col 0 = row index, col 1 = col index)
//         [1] values  [nnz] float
// IArgs:  [0] rows, [1] cols
// Outputs:[0] csrValues [nnz] float (same dtype as values)
//         [1] colIdx    [nnz] INT32
//         [2] rowPtr    [rows+1] INT32
//
// Entries are sorted by (row, col) in the output CSR representation.
//

#include <system/op_boilerplate.h>
#if NOT_EXCLUDED(OP_coo_to_csr)

#include <ops/declarable/headers/sparse.h>
#include <ops/declarable/helpers/sparse_coo.h>
#include <algorithm>
#include <vector>

namespace sd {
namespace ops {

CUSTOM_OP_IMPL(coo_to_csr, 2, 3, false, 0, 2) {
  auto cooIndices = INPUT_VARIABLE(0);
  auto cooValues  = INPUT_VARIABLE(1);

  const LongType rows = INT_ARG(0);
  const LongType cols = INT_ARG(1);

  REQUIRE_TRUE(cooIndices->rankOf() == 2, 0,
               "coo_to_csr: indices must be 2D [nnz, 2], got rank %d", cooIndices->rankOf());
  REQUIRE_TRUE(cooIndices->sizeAt(1) == 2, 0,
               "coo_to_csr: indices second dim must be 2, got %lld",
               (long long)cooIndices->sizeAt(1));
  REQUIRE_TRUE(cooValues->rankOf() == 1, 0,
               "coo_to_csr: values must be 1D, got rank %d", cooValues->rankOf());
  REQUIRE_TRUE(cooIndices->sizeAt(0) == cooValues->lengthOf(), 0,
               "coo_to_csr: indices rows (%lld) must match values length (%lld)",
               (long long)cooIndices->sizeAt(0), (long long)cooValues->lengthOf());

  auto csrValues = OUTPUT_VARIABLE(0);
  auto colIdx    = OUTPUT_VARIABLE(1);
  auto rowPtr    = OUTPUT_VARIABLE(2);

  sd::ops::helpers::cooToCsr(block.launchContext(), cooIndices, cooValues,
                              csrValues, colIdx, rowPtr, rows);

  return sd::Status::OK;
}

DECLARE_SHAPE_FN(coo_to_csr) {
  auto cooIndices = INPUT_VARIABLE(0);
  auto cooValues  = INPUT_VARIABLE(1);

  const LongType nnz  = cooValues->lengthOf();
  const LongType rows = INT_ARG(0);

  auto floatDtype = cooValues->dataType();
  auto idxDtype   = sd::DataType::INT32;

  // Build sorted list of (row, col) pairs and count distinct ones.
  // cooIndices has shape [nnz, 2]; linear index k*2+0 = row, k*2+1 = col.
  std::vector<std::pair<LongType, LongType>> pairs(static_cast<size_t>(nnz));
  for (LongType k = 0; k < nnz; ++k) {
    pairs[k].first  = cooIndices->e<LongType>(k * 2);
    pairs[k].second = cooIndices->e<LongType>(k * 2 + 1);
  }
  std::sort(pairs.begin(), pairs.end());

  LongType coalescedNnz = nnz > 0 ? 1 : 0;
  for (LongType k = 1; k < nnz; ++k) {
    if (pairs[k] != pairs[k - 1]) ++coalescedNnz;
  }

  // values and colIdx are sized to the coalesced count; rowPtr is always rows+1.
  auto cvShape = ConstantShapeHelper::getInstance().createShapeInfo(floatDtype, 'c', {coalescedNnz});
  auto ciShape = ConstantShapeHelper::getInstance().createShapeInfo(idxDtype,   'c', {coalescedNnz});
  auto rpShape = ConstantShapeHelper::getInstance().createShapeInfo(idxDtype,   'c', {rows + 1});

  return SHAPELIST(cvShape, ciShape, rpShape);
}

DECLARE_TYPES(coo_to_csr) {
  getOpDescriptor()
      ->setAllowedInputTypes(0,  {ALL_INTS})    // indices
      ->setAllowedInputTypes(1,  {ALL_FLOATS})  // values
      ->setAllowedOutputTypes(0, {ALL_FLOATS})  // csrValues
      ->setAllowedOutputTypes(1, {ALL_INTS})    // colIdx
      ->setAllowedOutputTypes(2, {ALL_INTS});   // rowPtr
}

}  // namespace ops
}  // namespace sd

#endif  // NOT_EXCLUDED(OP_coo_to_csr)

File: nd4j/nd4j-backends/nd4j-api-parent/nd4j-api/src/main/java/org/nd4j/linalg/api/ops/impl/sparse/CooToCsr.java
/*
 *  ******************************************************************************
 *  *
 *  *
 *  * This program and the accompanying materials are made available under the
 *  * terms of the Apache License, Version 2.0 which is available at
 *  * https://www.apache.org/licenses/LICENSE-2.0.
 *  *
 *  * See the NOTICE file distributed with this work for additional
 *  * information regarding copyright ownership.
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
 *  * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
 *  * License for the specific language governing permissions and limitations
 *  * under the License.
 *  *
 *  * SPDX-License-Identifier: Apache-2.0
 *  *****************************************************************************
 */

package org.nd4j.linalg.api.ops.impl.sparse;

import org.nd4j.autodiff.samediff.SDVariable;
import org.nd4j.autodiff.samediff.SameDiff;
import org.nd4j.linalg.api.buffer.DataType;
import org.nd4j.linalg.api.ndarray.INDArray;
import org.nd4j.linalg.api.ops.DynamicCustomOp;

import java.util.Arrays;
import java.util.List;
import java.util.ArrayList;

/**
 * Converts COO (Coordinate) sparse representation to CSR (Compressed Sparse Row) format.
 *
 * <p>C++ op name: {@code coo_to_csr}
 *
 * <p>Inputs:
 * <ol>
 *   <li>indices – 2D [nnz, 2], INT64 — row/col index pairs for each non-zero</li>
 *   <li>values  – 1D [nnz], floating dtype — the non-zero values</li>
 * </ol>
 * Integer args: rows, cols
 *
 * <p>Outputs:
 * <ol>
 *   <li>values  – 1D [nnz], same dtype as input values (index 1)</li>
 *   <li>colIdx  – 1D [nnz], INT32 — column index for each non-zero</li>
 *   <li>rowPtr  – 1D [rows+1], INT32 — row pointers</li>
 * </ol>
 */
public class CooToCsr extends DynamicCustomOp {

    /** No-arg constructor required for ImportClassMapping reflection. */
    public CooToCsr() {}

    /**
     * Construct from pre-built COO component arrays.
     *
     * @param indices 2D [nnz, 2] INT64 row/col index pairs
     * @param values  1D [nnz] non-zero values
     * @param rows    number of rows in the logical dense shape
     * @param cols    number of columns in the logical dense shape
     */
    public CooToCsr(INDArray indices, INDArray values, long rows, long cols) {
        super(new INDArray[]{indices, values}, null);
        addIArgument(rows, cols);
    }

    /**
     * SameDiff constructor.
     *
     * @param sd      SameDiff instance
     * @param indices SD variable for the COO index pairs
     * @param values  SD variable for the non-zero values
     * @param rows    number of rows
     * @param cols    number of columns
     */
    public CooToCsr(SameDiff sd, SDVariable indices, SDVariable values, long rows, long cols) {
        super(sd, new SDVariable[]{indices, values});
        addIArgument(rows, cols);
    }

    @Override
    public String opName() {
        return "coo_to_csr";
    }

    @Override
    public List<DataType> calculateOutputDataTypes(List<DataType> dataTypes) {
        // Output 0 (values) has same dtype as the values input (index 1)
        // Output 1 (colIdx) is INT32; output 2 (rowPtr) is INT32
        return Arrays.asList(dataTypes.get(1), DataType.INT32, DataType.INT32);
    }

    /**
     * Gradient w.r.t. the COO inputs.
     *
     * <p>Forward outputs: [0] csrValues [csr_nnz], [1] csrColIdx [csr_nnz], [2] csrRowPtr [rows+1]
     * <br>
     * {@code grads.get(0)} = upstream gradient for {@code csrValues} [csr_nnz] (float).
     * {@code grads.get(1)} and {@code grads.get(2)} correspond to structural INT outputs — ignored.
     *
     * <p>Returns: {@code [dCooIndices (zero), dCooValues]} — gradients for each forward input.
     */
    @Override
    public List<SDVariable> doDiff(List<SDVariable> grads) {
        SDVariable gradCsrValues = grads.get(0);       // [csr_nnz] float
        SDVariable cooIndices    = arg(0);              // [coo_nnz, 2] INT
        SDVariable[] fwdOuts     = outputVariables();
        SDVariable csrColIdx     = fwdOuts[1];          // [csr_nnz] INT32
        SDVariable csrRowPtr     = fwdOuts[2];          // [rows+1]  INT32

        long rows = iArguments.get(0);
        long cols = iArguments.get(1);

        SDVariable dCooValues = new CooToCsrBp(sameDiff,
                cooIndices, csrColIdx, csrRowPtr, gradCsrValues,
                rows, cols).outputVariables()[0];

        // Return gradients in forward-input order:
        //   [0] cooIndices — structural INT, zero gradient
        //   [1] cooValues  — dCooValues from the bp op
        List<SDVariable> result = new ArrayList<>();
        result.add(sameDiff.zerosLike(cooIndices));
        result.add(dCooValues);
        return result;
    }
}
